docs: various improvements (#1384)
neovim/nvim-lspconfig: doc/lspconfig.txt

Additionally, all options defined in can also be configured via the `setup {}`
call. See `:help vim.lsp.start_client()` for details.

The most commonly passed overrides to `setup {}` are:

- {capabilities} `table <string, string|table|bool|function>`

  a table which represents the neovim client capabilities. Useful for
  broadcasting to the server additional functionality (snippets, off-protocol
  features) provided by plugins.

- {cmd} `list[string]`

  a list where the first entry is the binary used to run the language server.
  Additional entries are passed as arguments.

- {handlers} `list[functions]`

  a list of handlers which override the function used to process a response
  from a given language server. Applied only to the server referenced by
  setup. See |lsp-handler|.

- {init_options} `table <string, string|table|bool>`

  a table passed during the initialization notification after launching
  a language server. Equivalent to the `initializationOptions` field found
  in `InitializeParams` in the LSP specification.

  See upstream server documentation for available initialization
  options.

- {on_attach} `function(client, bufnr)`

>   Callback invoked by neovim's built-in client when attaching a buffer to a

Author: > (Corporate Design would distinguish "Neovim" (the project) from "Nvim" (the editor) ;))

I know this is true but I still don't like it ha

>   Callback invoked by neovim's built-in client when attaching a buffer to a

Author: Anyways, I tried to split the Nvim/Neovim correctly

>   Callback invoked by neovim's built-in client when attaching a buffer to a

Reviewer: That's why I didn't actually suggest it ;)

I'll leave it to @justinmk to enforce the branding...

>   language server. Allows setting options (local or otherwise) conditioned

Reviewer: Add a forward reference to `EXAMPLE KEYBINDINGS` below?

  after a language server attaches, or overriding client options. Most
  commonly used for settings buffer local keybindings.

- {settings} `table <string, string|table|bool>`  

  The `settings` table is sent in `on_init` via a
  `workspace/didChangeConfiguration` notification from the neovim client to
  the language server. These settings allow a user to change optional runtime
  settings of the language server. 

  The lspconfig document generation pipeline converts the `package.json`
  distributed by many language servers that includes optional settings into
  documentation in `server_configurations.md`. As a worked example, to send
  the following:

  `pyright.disableLanguageServices`**: `boolean`
  `pyright.disableOrganizeImports`**: `boolean`

>   The nested keys need to be translated into a nested table and passed to

Author: Yes

==============================================================================
SERVER CONFIGURATIONS                                 *lspconfig-configurations*

While the `setup {}` function is the primary interface to lspconfig, for
servers for which there is not a configuration, it is necessary to define a
configuration directly. This can be useful if there is an outstanding PR that
is in review, or when developing a language server that is unavailable
publicly.

The `configs` module is a singleton where configs are defined. The schema for
validating using `vim.validate` is:
>
    configs.SERVER_NAME = {
      default_config = {'t'};
      on_new_config = {'f', true};
      on_attach = {'f', true};
      commands = {'t', true};
      docs = {'t', true};
    }
<
Where the structure of the docs table is as follows:
>
    docs = {
      description = {'s', true};
      default_config = {'t', true};
    }
<
`commands` is a map of `name:definition` key:value pairs, where `definition`
is a list whose first value is a function implementing the command and the
rest are either array values which will be formed into flags for the command

> or special keys like `description`. Example:

Author: Not sure I understand this one, I removed most "for a worked examples"

> or special keys like `description`. Example:

Reviewer: I meant that 'replace with "Example:"'  would have been my recommendation instead of 'remove "worked"' had I come across this first. Either is fine.

>
    require'lspconfig'.SERVER_NAME.setup{}

After you set `configs.SERVER_NAME` you can add arbitrary language-specific
functions to it if necessary.

Example:

>
    configs.texlab.buf_build = buf_build
<
==============================================================================
ADDING NEW SERVERS                                    *lspconfig-adding-servers*

The three steps for adding and enabling a a new server configuration are:

- load the lspconfig module (note, this is a stylistic choice)
>
   local lspconfig = require 'lspconfig'
<
- define the configuration

>
   local configs = require 'lspconfig/configs'

   -- Check if the config is already defined (useful when reloading this file)
   if not lspconfig.foo_lsp then
     configs.foo_lsp = {
       default_config = {
         cmd = {'/home/neovim/lua-language-server/run.sh'};
         filetypes = {'lua'};
         root_dir = function(fname)
           return lspconfig.util.find_git_ancestor(fname)
         end;
         settings = {};
       };
     }
   end

- call `setup()` to enable the FileType autocmd
>
   lspconfig.foo_lsp.setup{}
<
==============================================================================

> INDEX                                                          *lspconfig-index*

Reviewer: Does this need a separate section? This is just repeating two sentences from the top of the file -- if needed, I'd just add a tag to those? (Maybe renamed to `lspconfig-serverlist`, as "index" already has a different meaning in the context of LSP.)

Author: Justin suggested it (and the name). The main point was to have a direct helptag link to the section. I think a small amount of duplication is fine.

Reviewer: I understand, but can't you add the tag to the paragraph in the introduction? That strikes me as serving the same purpose without adding a second "source of truth".

The autogenerated documentation for all language servers is accessible in
`server_configurations.md`. Type `gf` on the filename from this document to jump
to the markdown
file.

==============================================================================

> ROOT DETECTION                                        *lspconfig-root-detection*

Reviewer: I think this section (and many users in general) would benefit from having a clear statement here at the beginning on the dual role of root directories for lspconfig:

1. the directory to pass to **servers** as `rootDirectory` (to do with whatever they do; this can include ignoring)
2. the directory for **lspconfig** to consider under the regime of a single server (i.e., anything under `root_directory` is attached to the same server instead of starting a new one).

This _is_ later explained under "single file support", but I believe it would be better to move that whole section to the beginning and treat it as an explanation of what "root directory" means in the context of lsp(config).

                                                            *lspconfig-root-dir*

lspconfig automatically launches language servers by defining a filetype
autocommand based on the `filetypes` specified in the default configuration of
each server, optionally overridable by the `filetypes` table passed to
`setup`.

This autocommand triggers a search from the current file position in the
filesystem hierarchy up to the top level directory of your filesystem. The
`root_dir` entry of each configuration is a function that returns true if the
current directory in this traversal matches a given root pattern.

The following utility functions are provided by lspconfig. Each function call
below, returns a function that takes as it's argument the current buffer path.

> Examples are provided for didactic purpose:

Author: These aren't the examples though, they include examples

> Examples are provided for didactic purpose:

Reviewer: Then just remove that sentence here...

- `util.root_pattern`: variadic function which takes multiple arguments, which
  corresponding to a different root pattern against which the contents of the
  current directory in the traversal will be matched against using
  `vim.fn.glob`. See `:help vim.fn.glob` for details.
>
   root_dir = util.root_pattern('pyproject.toml', 'requirements.txt')
<

> - `util.find_git_ancestor`: locates the first parent directory containing a

Reviewer: "a function that locates..."

  `.git` directory.
>
   root_dir = util.find_git_ancestor

> - `util.find_node_modules_ancestor`:  locates the first parent directory

Reviewer: same

  containing a `node_modules` directory.
>
   root_dir = util.find_node_modules_ancestor
<

> - `util.find_package_json_ancestor`:  locates the first parent directory

Reviewer: same

  containing a `package.json`.
>
   root_dir = util.find_json_ancestor
<
==============================================================================

> COMPOSITIONAL ROOT DIRECTORIES                      *lspconfig-root-composition*

Reviewer: -> "ADVANCED ROOT DIRECTORY DETECTION"

("Compositional" is awkward and not immediately understandable.)

The `root_dir` key in `config` and `setup` can hold any function of the form
>
  function custom_root_dir(filename, bufnr)
  returns nil | string
>
This allows for rich composition of root directory patterns which is necessary

> for some project structures. A worked example for kotlin:

Reviewer: "Example (for Kotlin):"

>
  local root_files = {
    'settings.gradle', -- Gradle (multi-project)
    'settings.gradle.kts', -- Gradle (multi-project)
    'build.xml', -- Ant
    'pom.xml', -- Maven
  }

  local fallback_root_files = {
    'build.gradle', -- Gradle
    'build.gradle.kts', -- Gradle
  }
  root_dir = function(fname)
    local primary = util.root_pattern(unpack(root_files))(fname)
    local fallback = util.root_pattern(unpack(fallback_root_files))(fname)
    return primary or fallback
  end
<
Browsing the source of the default configurations is recommended.

==============================================================================

> SINGLE FILE SUPPORT                              *lspconfig-single-file-support*

Reviewer: I don't much like this section as it's mainly complaining about lspconfig's own configurations (no matter where they come from), but if it'll be removed before 0.6 anyway, I won't argue about it.

Author: I'm going to keep it until I merge the PR, but even then I will still have it but change the wording to explain the limitations of the protocol

Reviewer: Yeah, hence my later suggestion to turn this into a section about the role of `root_directory` at the start of the section.

Language servers require each project to have a `root` in order to provide
many features that require cross-file indexing.

Some servers support not passing a root directory as a proxy for single file

> mode under which cross-file features may be degraded. While the built-in

Reviewer: or, in fact, are capable of creating their own project structure from the current file...

Author: 🙃

client does not require passing a root directory to a language server,
lspconfig requires detecting a project root as it is used internally to
determine whether to attach a given language server to a file or launch a new
instance.

As lspconfig configurations are community supported, some community members
have opted to have lspconfig detect a file's parent directory as a root.

As a consequence:

* Multi-file "scratch" projects will still allow for cross-file features.

> * If the default server root pattern does not special case $HOME and other

Reviewer: "does not treat `$HOME` and other large directories differently, ..."

  large directories, the server will show a delayed startup as it indexes all
  files under the current file's parent directory

This fallback pattern will be removed in a future version of lspconfig in
favor of more explicit handling on a per server basis.

==============================================================================
COMMANDS                                                    *lspconfig-commands*

- `:LspInfo` shows the status of active and configured language servers. Note
  that client id refers to the neovim RPC instance connected to a given
  language server.

The following commands support tab-completion for all arguments. All commands
that require a client id can either leverage tab-completion or the info
contained in `:LspInfo`:

> - `:LspStart <config_name>` Launch requested configured client. Will only

Reviewer: * "launches the requested..."
* "client, but only..."

  launch the server if it successfully resolves a root directory. Defaults to
  all *configured* servers matching the *current* *buffer's* filetype.

> - `:LspStop <client_id>` Stop the server with the given client id. Defaults to

Reviewer: "stops the server"

  stopping all servers active on the current buffer.

> - `:LspRestart <client_id>` Restart the client with the given client id. Will

Reviewer: "restarts the server ... and will"

  attempt to reattach to all previously attached buffers.

==============================================================================
EXAMPLE KEYBINDINGS                                    *lspconfig-keybindings*

lspconfig, and the core client, do not map any keybindings by default. The

> following is an example lua block which demonstrates how to leverage on-attach

Reviewer: "Lua block", "`on_attach`"

      buf_set_keymap('n', '<space>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
      buf_set_keymap('n', '<space>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
      buf_set_keymap('n', '<space>ca', '<cmd>lua vim.lsp.buf.code_action()<CR>', opts)
      buf_set_keymap('n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
      buf_set_keymap('n', '<space>e', '<cmd>lua vim.lsp.diagnostic.show_line_diagnostics()<CR>', opts)
      buf_set_keymap('n', '[d', '<cmd>lua vim.lsp.diagnostic.goto_prev()<CR>', opts)
      buf_set_keymap('n', ']d', '<cmd>lua vim.lsp.diagnostic.goto_next()<CR>', opts)
      buf_set_keymap('n', '<space>q', '<cmd>lua vim.lsp.diagnostic.set_loclist()<CR>', opts)
      buf_set_keymap('n', '<space>f', '<cmd>lua vim.lsp.buf.formatting()<CR>', opts)

    end

    -- Use a loop to conveniently call 'setup' on multiple servers and
    -- map buffer local keybindings when the language server attaches
    local servers = { 'pyright', 'rust_analyzer', 'tsserver' }
    for _, lsp in ipairs(servers) do
      nvim_lsp[lsp].setup {
        on_attach = on_attach,
        flags = {
          debounce_text_changes = 150,
        }
      }
    end
<

> Note: these keymappings are illustrative and override some infrequently used

Reviewer: "are meant for illustration"

default mappings.

==============================================================================
COMPLETION SUPPORT                                        *lspconfig-completion*

> Manual, triggered completion can be provided by neovim's built-in omnifunc.

Reviewer: "Manually triggered"

Reviewer: "Neovim"

See `:help omnifunc` for more details.

> For autocompletion, neovim does not offer built-in functionality at this time.

Reviewer: same

Consult the lspconfig wiki, which provides configuration examples for using a
completion plugin with the built-in client

==============================================================================
DEBUGGING                                                  *lspconfig-debugging*

While using language servers should be easy, debugging issues can be
challenging. It is first important to identify the source of the issue, which
is typically (in rough order):

- the language server itself
- a plugin
- overrides in a user configuration
- the built-in client in neovim core
- lspconfig

The first step in debugging is to test with a minimal configuration (such as
`../test/minimal_init.lua`). Historically, many users problems are due to
plugins or misconfiguration.

Should that fail, identifying which component is the culprit is challenging.
The following are the *only* categories of bugs that pertain to lspconfig.

> - The root directory inferred for your project is wrong. Or it should be

Reviewer: "wrong, or it..."

>   detected but is not based on a flaw in the lspconfig path utilities.

Reviewer: "but is not due to a flaw" ("bug"? "mistake"?)

- The server is launching, but you believe that the default settings,
  initialization options, or command arguments are suboptimal and should be
  replaced based on your understanding of the server documentation.

> All bugs pertaining to errors in core neovim client should be reported to the

Reviewer: "in Neovim's built-in client"

neovim core issue tracker. All bugs pertaining to plugins should be reported
to the respective plugin. All missing features in a language server should be
reported to the upstream language server issue tracker.

> For debugging lspconfig issues the most common mistakes users make are:

Reviewer: * comma before "the"

* maybe less accusatory: "the most common stumbling blocks" / "hurdles"

  - The language server is not installed or is otherwise not executable.
    lspconfig does not install language servers for you. Ensure the `cmd`
    defined in `server_configurations.md` is executable from the command
    line. If the absolute path to the binary is not supplied in `cmd`, ensure
    it is on your PATH.
  - No root detected. Lspconfig is built around the concept of projects. See
    |lspconfig-root-directory| for more details. Most of the time,
    initializing a git repo will suffice.

>   - Misconfiguration. Often times users will override `cmd`, `on_init`, or

Reviewer: "Oftentimes" or just "often"

    `handlers`. Ensure that you debug by using a stock configuration to ensure
    your customizations are not introducing issues.

|LspInfo| provides an overview of your active and configured language servers
which can be useful for debugging.

> Note, that it will not report any configuration changes applied in

Reviewer: remove comma

`on_new_config`.

==============================================================================
LOGGING                                                      *lspconfig-logging*

When debugging language servers, it is helpful to enable additional logging in

> the built-in client, specifically considering the RPC logs.

Reviewer: "Example:"

>
  vim.lsp.set_log_level 'trace'
  if vim.fn.has 'nvim-0.5.1' == 1 then
    require('vim.lsp.log').set_format_func(vim.inspect)
  end
<
Attempt to run the language server, and open the log with:

>
    :lua vim.cmd('e'..vim.lsp.get_log_path())
<

> Note, `ERROR` messages which include `stderr` only indicate that the log was

Reviewer: "Note that `ERROR` messages that include

> sent to `stderr`. Many servers counter-intuitively send innocuous messages

Reviewer: "harmless" ("warning", "information"?)

via stderr.

==============================================================================
SCOPE                                                         *lspconfig-scope*

lspconfig is a community effort to create default configurations that fit
within the scope of an official plugin for neovim. All features that are not
strictly providing default configurations for language servers will be removed

> from lspconfig in time. The power of the neovim lsp ecosystem is in the

Reviewer: capitalization: "Neovim", "LSP" (maybe do a `s/.../.../g` on those)

composability and flexibility of integrating multiple plugins which maximizes
user choice and freedom.

> lspconfig also opts to adhere strictly to the lsp specification, with some

Reviewer: On a related note, maybe use "`lspconfig`" generally? Just having this in lowercase even at the start of a sentence is a bit irritating. (Very small nit; feel free to ignore.)

==============================================================================
ROOT DETECTION                                        *lspconfig-root-detection*
                                                            *lspconfig-root-dir*

lspconfig automatically launches language servers by defining a filetype
autocommand based on the `filetypes` specified in the default configuration of
each server, optionally overridable by the `filetypes` table passed to
`setup`.

This autocommand triggers a search from the current file position in the
filesystem hierarchy up to the top level directory of your filesystem. The
`root_dir` entry of each configuration is a function that returns true if the
current directory in this traversal matches a given root pattern.

The following utility functions are provided by lspconfig. Each function call
below, returns a function that takes as its argument the current buffer path.

Examples are provided for didactic purpose:

- `util.root_pattern`: function which takes multiple arguments, each 
  corresponding to a different root pattern against which the contents of the
  current directory are matched using |vim.fin.glob()| while traversing up the

>   filesystem.

Reviewer: ...and add "Example:" here.

(The other functions are not "examples" but _the_ correct usage.)

Neovim supports the Language Server Protocol (LSP) via the built-in language
server client. LSP facilitates many features, some of which include:

- go-to-definition
- find-references
- hover
- completion
- rename
- format
- refactor

These features are implemented in neovim core, not lspconfig. See `:help lsp`
for more details.

NOTE: Feature availability depends on the implementation details of the
server. that servers may implement only a subset of these features. Always
consult the server documentation before filing a bug report on a missing
feature.

==============================================================================
QUICKSTART                                                *lspconfig-quickstart*

- ensure the server is installed and executable from the command line

- enable the server in your neovim configuration (lua example):
>
    require'lspconfig'.clangd.setup{}
<
- create a new project, ensure it contains a root marker which matches the
  server requirements specified in `servers.md`.

- open a file within that project, such as `main.c`.

==============================================================================

> SETUP                                                          *lspconfig-setup*

Reviewer 2: I feel like this makes it sound as if there's a function called `lspconfig.setup()`.

> The main entrypoint to enabling and customizing a language server is the

Reviewer 2: ```suggestion
The main entry-point to enabling and customizing a configuration of a language server is the
```

`setup {}` call. Using the defaults for a server built-into lspconfig is
simple:
>
    require'lspconfig'.clangd.setup{}
<
The available server names are listed in `./server.md` and match the server
name in `config.SERVER_NAME` defined in each configuration's source file.

The purpose of `setup{}` is to wrap the call to `vim.lsp.start_client`  which
is built into neovim core in an autocommand that calls `start_client` or
`vim.lsp.buf_attach_client()` depending on if the current file appears to be
in a project that contains a server previously started by lspconfig. See
|root_dir| for more details.

The `setup{}` function takes a table which contains a superset of the keys

> listed in `:help vim.lsp.start_client()` with the following unique entries:

Reviewer 2: I don't think this obvious when you have actually split some options into the override section and.

- {root_dir}

    `function(filename, bufnr)`

    Returns either a filepath (string) or nil. The language server will only
    start if the function returns a filepath.

>     If a root directory (string) is returned which is unique from any
>     previously returned root_dir, a new server will be spawned with that
>     root directory. See |lspconfig-root-directory| for more details

Reviewer 2: ```suggestion
    See |lspconfig-root-directory| for more details
```

- {name}

    `string`

    Defaults to the server's name (`clangd`, `pyright`, etc.).

- {filetypes}

    `list[string] | nil`

    Set of filetypes on which to attempt to resolve {root_dir}.

    May be empty or server may specify a default value.

- {autostart}

    `bool` (default: true)

    Controls if the `FileType` autocommand that launches a language server is
    enabled. If `false`, allows for deferring language servers until manually
    launched with `:LspStart` (|lspconfig-commands|).

- {on_new_config}

    `function(new_config, new_root_dir)`

    Function executed after a root directory is detected. This is used to
    modify the server configuration (including `cmd` itself). Most commonly,

>     this is used to inject additional arguments into `cmd`.

Reviewer 2: You need to add an explanation about this as well: https://github.com/neovim/nvim-lspconfig/blob/564490ba5c48314e4c194279fa72dba9874bc65a/lua/lspconfig/angularls.lua#L52.

Author: This isn't true, or at least, the way I wrote it originally was extremely misleading.

Reviewer 2: I'm not sure about the original use-case, but for `powershell_es` at least, I have to define both `on_new_config` and `cmd`, otherwise it wouldn't work.

>     If overriding `on_new_config`, ensure that you read the
>     `on_new_config` defined in the source file of the default configuration
>     in lspconfig. The original `on_new_config` for a given server should
>     be likely be a subset of your new override

Reviewer 2: ```suggestion
    Make sure to read the original pre-defined `on_new_config` for a given server if it exists before overriding it. As you'll likely want it to be a subset of your new override.
```

Author: I like my phrasing better 😆

Note: all entries passed to `setup{}` override the entry in the default
configuration. There is no composition.

==============================================================================
OVERRIDES VIA SETUP                                  *lspconfig-setup-overrides*

The previous keys allow for customizing a server without redefining the

> config. The following retains all default settings for gopls, but overrides
> the root pattern to only match `.git` directories:
> >
>   local lspconfig = require'lspconfig'
>   lspconfig.gopls.setup{
>     root_dir = lspconfig.util.root_pattern('.git');
>   }
> <

Reviewer 2: ```suggestion
config. The following retains all default settings for tsserver, but overrides
the filetypes to only allow it attach to `typesecript` files, instead of default preset
>
  local lspconfig = require'lspconfig'
  lspconfig.tsserver.setup{
    filetypes = {"typescript"}
  }
<
```

I think it's better to give an easier example here, given that you haven't explained `root_dir` or `root_pattern` yet.

Author: root_dir is what people will most common override, and I do briefly explain it.

                                                            *lspconfig-settings*
                                                                 *lspconfig-cmd*

The most commonly passed overrides to `setup` are:

- capabilities*
- cmd*
- filetypes
- handlers*
- init_options*
- on_attach*
- root_dir
- settings*

See `:help vim.lsp.start_client()` for details on all options shared with
`start_client()` which are annotated with an asterisk.

Briefly re-summarized:

- {capabilities} `table <string, string|table|bool|function>`

  a table which represents the neovim client capabilities. Useful for
  broadcasting to the server additional functionality (snippets, off-protocol
  features) provided by plugins.

- {cmd} `list[string]`

  a list where the first entry is the binary used to run the language server.
  Additional entries are passed as arguments.

- {filetypes} `list[string]`

  a list of filetypes for which to automatically launch the language server.
  Specific to lspconfig.

- {handlers} `list[functions]`

  a list of handlers which override the function used to process a response
  from a given language server. Applied only to the server referenced by

>   setup. 

Reviewer 2: ```suggestion
  setup. See |lsp-handler|
```

- {init_options} `table <string, string|table|bool>`

>   Passed as `initializationOptions` in `InitializeParams`.

Reviewer 2: ```suggestion
  Values to pass in the initialization request as `initializationOptions`. See `initialize` in the LSP spec.
```

>   See upstream server documentation for specific allowed initialization

Reviewer 2: ```suggestion
  Check the specific server documentation for any allowed initialization
```

  options.

- {on_attach} `function(client, bufnr)`

>   Function called by neovim's built-in client after attaching a buffer to a

Reviewer 2: ```suggestion
  Callback invoked by neovim's built-in client when attaching a buffer to a
```

  language server. Allows setting options (local or otherwise) conditioned
  after a language server attaches, or overriding client options. Most
  commonly used for settings buffer local keybindings.

- {root_dir} `function(filename, bufnr)`

  see |lspconfig-root-dir|

- {settings} `table <string, string|table|bool>`  

>   The `settings` table is sent in `on_init` via a
>   `workspace/didChangeConfiguration` notification from the neovim client to
>   the language server. These settings allow a user to change optional runtime
>   settings of the language server. 

Reviewer 2: ```suggestion
   Map with language server specific settings sent to the server via
  `workspace/didChangeConfiguration` notification by the neovim client.
```

  The lspconfig document generation pipeline converts the `package.json`
  distributed by many language servers that includes optional settings into
  documentation in `servers.md`. As a worked example, to send the following:

  `pyright.disableLanguageServices`**: `boolean`
  `pyright.disableOrganizeImports`**: `boolean`

  The nested keys need to be translated into a nested table and passed to
  settings as follows:
>
  require('lspconfig').pyright.setup{
    settings = {
      pyright = {
          disableLanguageServices = true,
          disableOrganizeImports  = true,
        }
    }
  }
<
  
  Note, the autogenerated settings occasionally include vscode specific
  settings. If a setting is not respected by a language server, consult
  upstream docuemntation.

==============================================================================

> OVERRIDING GLOBAL DEFAULTS                         *lspconfig-override-defaults*

Reviewer 2: ```suggestion
OVERRIDING GLOBAL DEFAULTS                         *lspconfig-override-global-defaults*
```

The global defaults for all servers can be overridden by extending the
default_config table. For a worked example, see below:

>
  local lspconfig = require'lspconfig'
  lspconfig.util.default_config = vim.tbl_extend(
    "force",
    lspconfig.util.default_config,
    {
      autostart = false,
      handlers = {
        ["window/logMessage"] = function(err, method, params, client_id)
            if params and params.type <= vim.lsp.protocol.MessageType.Log then
              vim.lsp.handlers["window/logMessage"](err, method, params, client_id)
            end
          end;
        ["window/showMessage"] = function(err, method, params, client_id)
            if params and params.type <= vim.lsp.protocol.MessageType.Warning.Error then
              vim.lsp.handlers["window/showMessage"](err, method, params, client_id)
            end
          end;
      }
    }
  )
<
`setup {}` can additionally override these defaults in subsequent calls.

==============================================================================
SERVER CONFIGURATIONS                                  *lspconfig-configuration*

> 

Reviewer 2: Shouldn't this be in the `setup` section?

Author: No, setup and the `config` interface are distinct

- load the lspconfig module (note, this is a stylistic choice)
>
   local lspconfig = require 'lspconfig'
<
- define the configuration

>
   local configs = require 'lspconfig/configs'

   -- Check if the config is already defined (useful when reloading this file)
   if not lspconfig.foo_lsp then
     configs.foo_lsp = {
       default_config = {
         cmd = {'/home/neovim/lua-language-server/run.sh'};
         filetypes = {'lua'};
         root_dir = function(fname)
           return lspconfig.util.find_git_ancestor(fname)
         end;
         settings = {};
       };
     }
   end

- call `setup()` to enable the FileType autocmd
>
   lspconfig.foo_lsp.setup{}
<

==============================================================================
COMMANDS                                                    *lspconfig-commands*

- `:LspInfo` shows the status of active and configured language servers. Note
  that client id refers to the neovim RPC instance connected to a given
  language server.

The following commands support tab-completion for all arguments. All commands
that require a client id can either leverage tab-completion or the info
contained in `:LspInfo`:

> - `:LspStart <config_name>` Launch requested configured client. Attempts to
>   detect a root directory for the given client and fails if not found.

Reviewer 2: ```suggestion
- `:LspStart <config_name>` Launch requested configured client.
 Will only successfully start if the command detects a root directory matching the current config. 

Pass `autostart = false` to your `.setup{}` call for a language server if you would like to launch clients solely with this command.
```

Author: I cut `Pass `autostart = false` to your `.setup{}` call for a language server if you would like to launch clients solely with this command.` explicitly

>   Defaults to all *configured* servers matching the current filetype.

Reviewer 2: Doesn't this actually mean "it will **only** work for a matching the current buffer `filetype` ". It's probably the same meaning, but I think it could be a bit more obvious that nothing will happen if there are no matching servers, even if you try to force it. e.g. `:LspStart clangd` in a javascript file.

- `:LspStop <client_id>` Stop the server with the given client id. Defaults to
  stopping all servers active on the current buffer.
- `:LspRestart <client_id>` Restart the client with the given client id. Will
  attempt to reattach to all previously attached buffers.

==============================================================================
ROOT DETECTION                                        *lspconfig-root-detection*
                                                            *lspconfig-root-dir*

> lspconfig automatically launches language servers by defining a filetype

Reviewer 2: Maybe an explanation/refresher first that this an LSP requirement
`:h vim.lsp.start_client()`
> Directory where the LSP server will base its rootUri on initialization.

Author: It's not required by the spec, it's assumed by many language servers.

In fact we send nil in #1385

- {on_attach} `function(client, bufnr)`

  Callback invoked by Nvim's built-in client when attaching a buffer to a
  language server. Often used to set Nvim (buffer or global) options or to
  override the Nvim client properties (`resolved_capabilities`) after a
  language server attaches. Most commonly used for settings buffer
  local keybindings. See |lspconfig-keybindings| for a usage example.

- {settings} `table <string, string|table|bool>`  

  The `settings` table is sent in `on_init` via a
  `workspace/didChangeConfiguration` notification from the Nvim client to
  the language server. These settings allow a user to change optional runtime
  settings of the language server. 

  The scripts that automatically generate `server_configurations.md` convert 

>   the `package.json` referened in many a server's configuration source file

Reviewer: "referenced"

Reviewer 2: How many "scripts" are there? isn't it just the one? Also, wouldn't it be better to make it clear that this is more or less specific to servers implemented in typescript/javascript?

Author: it's not tho

  into a list of optional settings listed in `server_configurations.md`.

  As an example, to set the following settings found in the pyright
  documentation:

  `pyright.disableLanguageServices`: `boolean`
  `pyright.disableOrganizeImports`: `boolean`

>   nested keys need to be translated into a nested table and passed to

Reviewer: Capitalize "Nested"?

current directory in this traversal matches a given root pattern.

The following utility functions are provided by lspconfig. Each function call

> below, returns a function that takes as its argument the current buffer path.

Reviewer: forgot to remove comma ;)

As a consequence:

* Multi-file "scratch" projects will still allow for cross-file features.
* The server treats `$HOME` the same as any other large directory. The server

>   will index all files unde the project root which can delay startup.

Reviewer: "under"
